log/loki.go

```diff
@@ -0,0 +1,486 @@
+			// of allocation. Combined with the cutoff part and directly pushing in the final data
+			// type this can be really a lot faster and to use a lot less memory
+			labels := make(map[string]string, len(entry.Data)+1)
+			for k, v := range entry.Data {
+				labels[k] = fmt.Sprint(v) // TODO optimize ?
+			}
+			for _, params := range h.labels {
+				labels[params[0]] = params[1]
+			}
+			labels["level"] = entry.Level.String()
+			// have the cutoff here ?
+			// if we cutoff here we can cut somewhat on the backbuffers and optimize the inserting
+			// in/creating of the final Streams that we push
+			msgs[count] = tmpMsg{
+				labels: labels,
+				msg:    entry.Message,
+				t:      entry.Time.UnixNano(),
+			}
+			count++
+		case <-ticker.C:
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- h.pushPeriod / 2
+			<-ch
+		case <-h.ctx.Done():
+			ch := make(chan time.Duration)
+			pushCh <- ch
+			ch <- 0
+			<-ch
+			return
+		}
+	}
+}
+
+func sortAndSplitMsgs(msgs []tmpMsg, bufferTime time.Duration) int {
+	if len(msgs) == 0 {
+		return 0
+	}
+
+	// TODO using time.Before was giving a lot of out of order, but even now, there are some, if the
+	// limit is big enough ...
+	sort.Slice(msgs, func(i, j int) bool {
+		return msgs[i].t < msgs[j].t
+	})
+
+	// We can technically cutoff during the addMsg phase, which will be even better if we sort after
+	// it as well ... maybe
+	cutOff := time.Now().Add(-bufferTime).UnixNano() // probably better to be configurable
+
+	cutOffIndex := sort.Search(len(msgs), func(i int) bool {
+		return !(msgs[i].t < cutOff)
+	})
+	return cutOffIndex
+}
+
+func (h *lokiHook) createPushMessage(msgs []tmpMsg, cutOffIndex, dropped int) *lokiPushMessage {
+	strms := new(lokiPushMessage)
+	strms.msgMaxSize = h.msgMaxSize
+	for _, msg := range msgs[:cutOffIndex] {
+		strms.add(msg)
+	}
+	if dropped != 0 {
+		labels := make(map[string]string, 2+len(h.labels))
+		labels["level"] = logrus.WarnLevel.String()
+		labels["dropped"] = strconv.Itoa(dropped)
+		for _, params := range h.labels {
+			labels[params[0]] = params[1]
+		}
+
+		msg := tmpMsg{
+			labels: labels,
+			msg: fmt.Sprintf("k6 dropped some packages because they were above the limit of %d/%s",
+				h.limit, h.pushPeriod),
+			t: msgs[cutOffIndex-1].t,
+		}
+		strms.add(msg)
+	}
+	return strms
+}
+
+func (h *lokiHook) push(b bytes.Buffer) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	body := b.Bytes()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", h.addr, &b)
+	if err != nil {
+		return err
+	}
+	req.GetBody = func() (io.ReadCloser, error) {
+		return ioutil.NopCloser(bytes.NewBuffer(body)), nil
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := h.client.Do(req)
+
+	if res != nil {
+		if res.StatusCode == 400 {
+			r, _ := ioutil.ReadAll(res.Body) // maybe limit it to something like the first 1000 characters?
+			fmt.Println("Got 400 from loki: " + string(r))
+		}
```

Comment:
This print statement probably has to be removed and converted to an error result of the method. And it seems appropriate to make the check be for any non-2xx status, since I imagine a 500 or a 401 status wouldn't be any better than 400...

Reply:
yeah, I mostly added this as 400 is what loki returns everytime it doesn't like something ...
Including if (even after all the code about it) k6 pushes a message from before the latest one ...